chainntnfs/interface_test.go

```diff
@@ -1310,10 +1310,308 @@ func testReorgConf(miner *rpctest.Harness, notifier chainntnfs.ChainNotifier,
 	}
 }
 
+// testCatchUpClientOnMissedBlocks tests the case of multiple registered client
+// receiving historical block epoch notifications due to their best known block
+// being out of date.
+func testCatchUpClientOnMissedBlocks(miner *rpctest.Harness,
+	notifier chainntnfs.TestChainNotifier, t *testing.T) {
+
+	const numBlocks = 10
+	const numClients = 5
+	var wg sync.WaitGroup
+
+	outdatedHash, outdatedHeight, err := miner.Node.GetBestBlock()
+	if err != nil {
+		t.Fatalf("unable to retrieve current height: %v", err)
+	}
+
+	// Generate numBlocks blocks which clients should receive historical
+	// notifications for.
+	if _, err := miner.Node.Generate(numBlocks); err != nil {
+		t.Fatalf("unable to generate blocks: %v", err)
+	}
+
+	// We want to ensure that when a client registers for block notifications,
+	// the notifier's best block is at the tip of the chain. If it isn't, the
+	// client may not receive all historical notifications.
+	bestHash, bestHeight, err := miner.Node.GetBestBlock()
+	if err != nil {
+		t.Fatalf("unable to retrieve current height: %v", err)
+	}
+
+	bestBlock := chainntnfs.BlockEpoch{Height: bestHeight, Hash: bestHash}
+	if err := notifier.UnsafeStart(bestBlock); err != nil {
+		t.Fatalf("Unable to unsafe start the notifier: %v", err)
+	}
+
+	// Create numClients clients whose best known block is 10 blocks behind
+	// the tip of the chain. We expect each client to receive numBlocks
+	// notifications, 1 for each block  they're behind.
+	clients := make([]*chainntnfs.BlockEpochEvent, 0, numClients)
+	outdatedBlock := &chainntnfs.BlockEpoch{
+		Height: outdatedHeight, Hash: outdatedHash,
+	}
+	for i := 0; i < numClients; i++ {
+		epochClient, err := notifier.RegisterBlockEpochNtfn(outdatedBlock)
+		if err != nil {
+			t.Fatalf("unable to register for epoch notification: %v", err)
+		}
+		clients = append(clients, epochClient)
+	}
+	for expectedHeight := outdatedHeight + 1; expectedHeight <=
+		bestHeight; expectedHeight++ {
+
+		for _, epochClient := range clients {
+			select {
+			case block := <-epochClient.Epochs:
+				if block.Height != expectedHeight {
+					t.Fatalf("received block of height: %d, "+
+						"expected: %d", block.Height,
+						expectedHeight)
+				}
+			case <-time.After(20 * time.Second):
+				t.Fatalf("did not receive historical notification")
+			}
+
+		}
+	}
+
+	// Finally, ensure that an extra block notification wasn't received.
+	for _, epochClient := range clients {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			select {
+			case <-epochClient.Epochs:
+				t.Fatalf("received unexpected block notification")
+			case <-time.After(5 * time.Second):
+			}
+		}()
```

Comment:
we'll need to pass in `epochClient` as an argument to the goroutine, to avoid capturing the local variable. otherwise we may have data races. will need to be done on the other test cases as well

Reply:
fixed in all tests :)

Reply:
heh, looks like there's still anotherrrr race condition. Unfortunately there's actually a race condition in `Fatalf`, so not your fault :P

was able to get rid of the race conditions doing this:
```golang
        // Finally, ensure that an extra block notification wasn't received.
        anyExtras := make(chan struct{}, len(clients))                                                                                                         
        for _, epochClient := range clients {                                                                                                                  
                wg.Add(1)
                go func(epochClient *chainntnfs.BlockEpochEvent) {                                                                                             
                        defer wg.Done()                                                                                                                        
                        select {
                        case <-epochClient.Epochs:
                                anyExtras <- struct{}{}
                        case <-time.After(5 * time.Second):                                                                                                    
                        }
                }(epochClient)                                                                                                                                 
        }                                                                                                                                                      

        wg.Wait()                                                                                                                                              
        close(anyExtras)                                                                                                                                       

        var extraCount int                                                                                                                                     
        for range anyExtras {                                                                                                                                  
                extraCount++                                                                                                                                   
        }                                                                                                                                                      

        if extraCount > 0 {                                                                                                                                    
                t.Fatalf("received %d unexpected block notification", extraCount)                                                                              
        }  
```

Reply:
ah i see it on travis but i thought i didn't see it locally when running `-race`.. let me know if you know of a reason for that, otherwise i may just be mistaken. 

implemented this change, tysm :)